src/high-level/constructors.lisp

```diff
@@ -0,0 +1,173 @@
+
+If TYPE is not specified then it is inferred from the element type of ARRAY.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let* (;; TODO: This gets the type of an array from the array, not
+           ;; by inspecting the elements. This can cause issues when
+           ;; the element-type of the array is T
+           (element-type
+             (if (null type)
+                 (array-element-type array)
+                 type))
+           (tensor-class (infer-tensor-type element-type shape nil)))
+      (adjust-array array (list (reduce #'* shape)) :element-type element-type)
+      (make-tensor tensor-class shape
+                   :storage array
+                   :layout layout))))
+
+(defun from-list (list shape &key type layout (input-layout :row-major))
+  "Create a tensor with the elements of LIST, placing in layout INPUT-LAYOUT
+
+If INPUT-LAYOUT is not specified then row-major is assumed.
+
+If TYPE is not specified then it is inferred from the type of the first element of LIST.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape)
+       (assertion (cl:= (length list) (reduce #'* shape))))
+    (let ((tensor-class (infer-tensor-type type shape (first list))))
+      (let ((tensor (make-tensor tensor-class shape :layout layout)))
+        (into!
+         (lambda (&rest pos)
+           (nth
+            (if (eql input-layout :row-major)
+                (row-major-index pos shape)
+                (column-major-index pos shape))
+            list))
+         tensor)))))
+
+(defun from-diag (list &key (order 2) type layout)
+  "Create a tensor from a list, placing along the diagonal
+
+If ORDER is specified then the tensor will be of that order, otherwise 2 is assumed.
+If TYPE is not specified then it is inferred from the type of the first element of LIST.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (let* ((shape (fixnum-to-shape (length list) order))
+         (tensor-class (infer-tensor-type type shape (first list)))
+         (tensor (make-tensor tensor-class shape :layout layout)))
+    (loop :for i :below (first shape)
+          :do (setf (tref tensor i i) (pop list)))
+    tensor))
+
+;;; Constructors for convenience
+
+(defun zeros (shape &key (type +default-tensor-type+) layout)
+  "Create a tensor with the specified SHAPE of zeros
+
+If TYPE is not specified then +DEFAULT-TENSOR-TYPE+ is used.
+LAYOUT specifies the internal storage represenation ordering of the returned tensor.
```

Comment:
```suggestion
LAYOUT specifies the internal storage representation ordering of the returned tensor.
```

Reply:
Turns out copying and pasting the same docstring can lead to mistakes when you can't spell